pkg/proxy/handlers/spacelister.go

```diff
@@ -81,68 +150,98 @@ func (s *SpaceLister) ListUserWorkspaces(ctx echo.Context) ([]toolchainv1alpha1.
-			log.Errorf(nil, err, "unable to get space", "space", spaceName)
+			log.Errorf(nil, err, "unable to get space", "space", spacebinding.Labels[toolchainv1alpha1.SpaceBindingSpaceLabelKey])
 			continue
 		}
+		workspace := createCommonWorkspace(signupName, space, spacebinding)
+		workspaces = append(workspaces, *workspace)
+	}
+	return workspaces
+}
 
-		// TODO right now we get SpaceCreatorLabelKey but should get owner from Space once it's implemented
-		ownerName = space.Labels[toolchainv1alpha1.SpaceCreatorLabelKey]
+func createCommonWorkspace(signupName string, space *toolchainv1alpha1.Space, spaceBinding *toolchainv1alpha1.SpaceBinding, wsAdditionalOptions ...commonproxy.WorkspaceOption) *toolchainv1alpha1.Workspace {
+	// TODO right now we get SpaceCreatorLabelKey but should get owner from Space once it's implemented
+	ownerName := space.Labels[toolchainv1alpha1.SpaceCreatorLabelKey]
 
-		wsOptions := []commonproxy.WorkspaceOption{
-			commonproxy.WithNamespaces(space.Status.ProvisionedNamespaces),
-			commonproxy.WithOwner(ownerName),
-			commonproxy.WithRole(spaceBinding.Spec.SpaceRole),
-			commonproxy.WithObjectMetaFrom(space.ObjectMeta),
-		}
-		// set the workspace type to "home" to indicate it is the user's home space
-		// TODO set home type based on UserSignup.Status.HomeSpace once it's implemented
-		if ownerName == signupName {
-			wsOptions = append(wsOptions, commonproxy.WithType("home"))
-		}
+	wsOptions := []commonproxy.WorkspaceOption{
+		commonproxy.WithNamespaces(space.Status.ProvisionedNamespaces),
+		commonproxy.WithOwner(ownerName),
+		commonproxy.WithRole(spaceBinding.Spec.SpaceRole),
+		commonproxy.WithObjectMetaFrom(space.ObjectMeta),
+	}
+	// set the workspace type to "home" to indicate it is the user's home space
+	// TODO set home type based on UserSignup.Status.HomeSpace once it's implemented
+	if ownerName == signupName {
+		wsOptions = append(wsOptions, commonproxy.WithType("home"))
+	}
+	wsOptions = append(wsOptions, wsAdditionalOptions...)
 
-		workspace := commonproxy.NewWorkspace(spaceName, wsOptions...)
-		workspaces = append(workspaces, *workspace)
+	workspace := commonproxy.NewWorkspace(space.GetName(), wsOptions...)
+	return workspace
+}
+
+func (s *SpaceLister) getSpace(spaceBinding *toolchainv1alpha1.SpaceBinding) (*toolchainv1alpha1.Space, error) {
+	spaceName := spaceBinding.Labels[toolchainv1alpha1.SpaceBindingSpaceLabelKey]
+	if spaceName == "" { // space may not be initialized
+		// log error and continue so that the api behaves in a best effort manner
+		return nil, fmt.Errorf("spacebinding has no '%s' label", toolchainv1alpha1.SpaceBindingSpaceLabelKey)
 	}
-	return workspaces
+	return s.GetInformerServiceFunc().GetSpace(spaceName)
+}
+
+func getRolesFromNSTemplateTier(nstemplatetier *toolchainv1alpha1.NSTemplateTier) []string {
+	roles := make([]string, len(nstemplatetier.Spec.SpaceRoles))
+	i := 0
+	for k := range nstemplatetier.Spec.SpaceRoles {
+		roles[i] = k
+		i++
+	}
```

Comment:
Shouldn't it be this instead:
```suggestion
	for i, k := range nstemplatetier.Spec.SpaceRoles {
		roles[i] = k
	}
```

Reply:
Well, actually I now realized why I had to do it this way. The `SpaceRoles` indexes are strings (name of the role) and not integers.

Reply:
Ah! It's a map. Not an array.

Reply:
correct!